rust/lance/src/dataset/optimize.rs

```diff
@@ -498,15 +554,204 @@ pub struct RewriteResult {
     pub read_version: u64,
     /// The original fragments being replaced
     pub original_fragments: Vec<Fragment>,
+    pub remapped_indices: Vec<RemappedIndex>,
+}
+
+#[pin_project]
+struct RowIdCapture {
+    row_ids: Arc<RwLock<RoaringTreemap>>,
+    #[pin]
+    target: SendableRecordBatchStream,
+}
+
+impl RowIdCapture {
+    fn new_pinned(
+        target: SendableRecordBatchStream,
+        row_ids: Arc<RwLock<RoaringTreemap>>,
+    ) -> Pin<Box<Self>> {
+        Box::pin(Self { row_ids, target })
+    }
+
+    fn extract_row_ids(
+        row_ids: &mut RoaringTreemap,
+        batch: DFResult<RecordBatch>,
+    ) -> DFResult<RecordBatch> {
+        let batch = batch?;
+
+        let (row_id_idx, _) = batch
+            .schema()
+            .column_with_name(ROW_ID)
+            .expect("Received a batch without row ids");
+        let row_ids_arr = batch.column(row_id_idx);
+        let row_ids_itr = row_ids_arr
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .expect(&format!(
+                "Row ids had an unexpected type: {}",
+                row_ids_arr.data_type()
+            ))
+            .iter()
+            .map(|x| x.expect("Null row id encountered"));
+        row_ids.append(row_ids_itr).map_err(|err| {
+            datafusion::error::DataFusionError::Execution(format!(
+                "Row ids did not arrive in sorted order: {}",
+                err
+            ))
+        })?;
+        let non_row_ids_cols = (0..batch.num_columns())
+            .filter(|col| *col != row_id_idx)
+            .collect::<Vec<_>>();
+        Ok(batch.project(&non_row_ids_cols)?)
+    }
+}
+
+impl Stream for RowIdCapture {
+    type Item = DFResult<RecordBatch>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        let batch = ready!(this.target.poll_next_unpin(cx));
+        let mut row_ids = this.row_ids.write().unwrap();
+        match batch {
+            Some(batch) => {
+                let extracted = RowIdCapture::extract_row_ids(&mut row_ids, batch);
+                Poll::Ready(Some(extracted))
+            }
+            None => Poll::Ready(None),
+        }
+    }
+}
+
+impl datafusion::physical_plan::RecordBatchStream for RowIdCapture {
+    fn schema(&self) -> SchemaRef {
+        self.target.schema()
+    }
+}
+
+struct MissingIds<'a, I: Iterator<Item = u64>> {
```

Comment:
not really critical, but: Why make `MissingIds` generic over iterator? It seems like we could likely get some better performance by taking advantage of the unique structure of treemap. For example, we could call [bitmaps](https://docs.rs/roaring/0.10.2/roaring/treemap/struct.RoaringTreemap.html#method.bitmaps) to get an iterator over the fragment_id and local row ids Roaring bitmaps. And then on the local bitmaps, we can go in chunks and call [contains_range](https://docs.rs/roaring/0.10.2/roaring/bitmap/struct.RoaringBitmap.html#method.contains_range) to see if we move on early. That would be optimal for the common case where we have sparse deleted rows (especially given our threshold is only 10%).

Reply:
Fair point.  I didn't even consider this really.  Compaction is already pretty heavily O(# of modified rows) so I figured one more pass through wouldn't hurt.  I'll leave this for a potential perf improvement down the road.

Reply:
Added to #1369